### ISSUE TITLE
Merge unmerged branches from code-dot-org/craft

### ISSUE THREAD

**Issue by nobody5050** [craft-devs/craft-engine#20]
There are numerous pull requests open on code-dot-org/craft which introduce new features to the game. Provided they function, they should be merged over here with credit given to the original author of the commits.

**Comment by Alvarito050506** [comment]
Never thought of this, but it would be really nice.

**Comment by Alvarito050506** [comment]
Ok so, code-dot-org/craft#464 is the only "ready" PR that is actually useful here in my opinion.

There are lots of issues which might be useful to solve here, though. For example (not referencing them to avoid spam):

### Cleanup
 + #555
 + #548
 + #535
 + #515 (I'm not sure about this one)
 + #477

### Bugs
 + #462 (Really bad bug)
 + #456 (Not just a funny number, but a bug I have noticed)

### Enhancements
 + #460 (Depends on #459/#464)

**Comment by nobody5050** [comment]
I agree. We should also work on figuring out how their api works, or worked. In the screenshot in their readme the api is exposed using functions like `moveForward()`. Being able to access functions like that would lead to much better level design.

**Comment by Alvarito050506** [comment]
Actually, I'm guessing that the textarea in their README is just "binded" to the API via the global scope, just like our CodeMirror/demo is binded via the `api` object. [Here](https://github.com/craft-devs/craft-engine/blob/mastter/src/js/game/API/CodeOrgAPI.js#L85) is the definition of `api.moveFoward`, for example.

**Comment by Alvarito050506** [comment]
> Being able to access functions like that would lead to much better level design.

About this, I think it might lead to much _faster_ level design, but for "production" (non-demos), the API shouldn't be exposed, or at least encapsulated, in my opinion.

**Comment by StealthHydra179** [comment]
I think there should be 2 builds? 1 for level designers, and another for non level designers? @Alvarito050506

**Comment by nobody5050** [comment]
Maybe, although just hiding the code editor and console by default and making them a “dev mode” would work too

**Comment by StealthHydra179** [comment]
But isn't the code mode part of code-org default like that you can move? I'm not sure.

**Comment by nobody5050** [comment]
Yes but it’s accessible only via blackly, the direct api isn’t actually exposed to the end user. In fact some levels have api code in them that doesn’t show in the blackly interface
